File: server/app/schemas/analysis.py
from typing import Literal, Optional
from pydantic import BaseModel, ConfigDict, Field

from app.services.extractors import DesignSignals


class AnalyzeRequest(BaseModel):
    url: str


class CamelModel(BaseModel):
    model_config = ConfigDict(populate_by_name=True)


class ScoredSection(CamelModel):
    score: float | None = None
    suggestions: list[str] = Field(default_factory=list)


class ExperienceSignals(CamelModel):
    has_primary_cta_above_fold: bool = False
    hero_text_length: int = 0
    has_clear_value_prop: bool = False


class ContentAnalysis(ScoredSection):
    title: str | None = None
    meta_description: str | None = Field(
        default=None,
        serialization_alias="metaDescription",
    )
    experience_signals: Optional["ExperienceSignals"] = Field(
        default=None,
        serialization_alias="experienceSignals",
    )


class StructureAnalysis(ScoredSection):
    h1_headings: list[str] = Field(
        default_factory=list,
        serialization_alias="h1Headings",
    )
    h2_headings: list[str] = Field(
        default_factory=list,
        serialization_alias="h2Headings",
    )


class DesignAnalysis(ScoredSection):
    signals: DesignSignals


class UXAnalysis(ScoredSection):
    pass


class SEOAnalysis(ScoredSection):
    pass


class AnalysisData(CamelModel):
    url: str
    content: ContentAnalysis
    structure: StructureAnalysis
    design: DesignAnalysis
    ux: UXAnalysis
    seo: SEOAnalysis
    overall_score: float | None = Field(
        default=None,
        serialization_alias="overallScore",
    )


class AnalyzeResponse(BaseModel):
    status: Literal["success", "error", "partial"]
    data: AnalysisData | None = None
    error: str | None = None

File: client/src/types/analysis.ts
export interface DesignSignals {
  has_animations: boolean;
  animation_keyword_count: number;
  has_animation_library: boolean;
  detected_libraries: string[];
  image_count: number;
  video_count: number;
  has_media: boolean;
  // Extended signals (Phase 2)
  animation_density: "low" | "medium" | "high" | "none" | null;
  has_hover_feedback: boolean | null;
  has_scroll_animations: boolean | null;
}

export interface ExperienceSignals {
  has_primary_cta_above_fold: boolean | null;
  hero_text_length: number | null;
  has_clear_value_prop: boolean | null;
}

export interface ScoredSection {
  score: number | null;
  suggestions: string[];
}

export interface ContentAnalysis extends ScoredSection {
  title: string | null;
  metaDescription: string | null;
  // Optional hero/CTA signals forwarded from the experience extractor
  experienceSignals?: ExperienceSignals | null;
}

export interface StructureAnalysis extends ScoredSection {
  h1Headings: string[];
  h2Headings: string[];
}

export interface DesignAnalysis extends ScoredSection {
  signals: DesignSignals;
}

export type UXAnalysis = ScoredSection;

export type SEOAnalysis = ScoredSection;

export interface AnalysisData {
  url: string;
  content: ContentAnalysis;
  structure: StructureAnalysis;
  design: DesignAnalysis;
  ux: UXAnalysis;
  seo: SEOAnalysis;
  overallScore: number | null;
}

export interface AnalyzeResponse {
  status: "success" | "error" | "partial";
  data: AnalysisData | null;
  error: string | null;
}

export const SECTION_WEIGHTS = {
  content: 30,
  ux: 30,
  design: 25,
  structure: 10,
  seo: 5,
} as const;
